lib/problems.rb

```diff
@@ -1,13 +1,43 @@
 require_relative './stack.rb'
 
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n)
+# Space Complexity: O(1)
 def balanced(string)
-  raise NotImplementedError, "Not implemented yet"
+  return true if string.empty?
+  
+  stack = Stack.new
+  openers = ["(", "[", "{"]
+  valid_pairs = {"(" => ")", "[" => "]", "{" => "}"}
+  
+  string.split.each do |char|
+    if openers.include?(char) 
+      stack.push(char) 
+    else 
+      popped = stack.pop
+      return false if valid_pairs[popped] != char
+    end
+  end
+  
+  return true if stack.empty?
+  return false
 end
 
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n)
+# Space Complexity: O(1)
 def evaluate_postfix(postfix_expression)
```

Comment:
👍 , Nice work, although your space complexity is O(n) since you are building a stack.